src/main/java/com/google/cloud/tools/app/impl/cloudsdk/internal/process/DefaultProcessRunner.java

```diff
@@ -124,49 +133,79 @@ public void setStdErrLineListener(ProcessOutputLineListener stdErrLineListener)
   /**
    * Sets the subprocess exit listener for collecting the exit code of the subprocess.
    *
-   * @param exitListener Can be nul.
+   * @param exitListener Can be null.
    */
   public void setExitListener(ProcessExitListener exitListener) {
     this.exitListener = exitListener;
   }
 
   /**
-   * @return The process that is currently executing or was the last one to be started using the run
-   *     method.
+   * The process that is currently executing or was the last one to be started using the run
+   * method.
    */
   public Process getProcess() {
     return this.process;
   }
 
   private void handleErrOut(final Process process) {
-    if (stdErrLineListener != null) {
-      final Scanner stdErr = new Scanner(process.getErrorStream());
-      Thread stdErrThread = new Thread("standard-err") {
-        public void run() {
-          while (stdErr.hasNextLine() && !Thread.interrupted()) {
-            String line = stdErr.nextLine();
-            stdErrLineListener.outputLine(line);
-          }
+    final Scanner stdErr = new Scanner(process.getErrorStream());
+    Thread stdErrThread = new Thread("standard-err") {
+      public void run() {
+        while (stdErr.hasNextLine() && !Thread.interrupted()) {
+          String line = stdErr.nextLine();
+          consumeLine(line, true);
         }
-      };
-      stdErrThread.setDaemon(true);
-      stdErrThread.start();
-    }
+      }
+    };
+    stdErrThread.setDaemon(true);
+    stdErrThread.start();
   }
 
   private void handleStdOut(final Process process) {
-    if (stdOutLineListener != null) {
-      final Scanner stdOut = new Scanner(process.getInputStream());
-      Thread stdOutThread = new Thread("standard-out") {
-        public void run() {
-          while (stdOut.hasNextLine() && !Thread.interrupted()) {
-            String line = stdOut.nextLine();
-            stdOutLineListener.outputLine(line);
-          }
+    final Scanner stdOut = new Scanner(process.getInputStream());
+    Thread stdOutThread = new Thread("standard-out") {
+      public void run() {
+        while (stdOut.hasNextLine() && !Thread.interrupted()) {
+          String line = stdOut.nextLine();
+          consumeLine(line, false);
         }
-      };
-      stdOutThread.setDaemon(true);
-      stdOutThread.start();
+      }
+    };
+    stdOutThread.setDaemon(true);
+    stdOutThread.start();
+  }
+
+  private void handleWaitSuccess() throws InterruptedException, ProcessRunnerException {
+    try {
+      if (async && waitSuccessMessage != null && waitSuccessTimeoutSeconds != 0) {
+        log.info("Waiting " + waitSuccessTimeoutSeconds
+            + " seconds for the operation to succeed...");
+        if (!waitSuccessLatch.await(waitSuccessTimeoutSeconds, TimeUnit.SECONDS)) {
```

Comment:
even for sync runs this shouldn't block because this would be called after the process exits